### ISSUE TITLE
The maximum number of active connections for UID 0 has been reached

### ISSUE THREAD

**Issue by wes8ty** [python-sdbus/python-sdbus#84]
Hoping someone can help. I am using the `sdbus-networkmanager` to monitor the connection status in a cron task. It seemed over time that the number of connections reached a max threshold (dbus daemon). In order to recover, I have to restart NetworkManager to get it to work again. I am trying to figure out how the sdbus library opens/closes connections via the dbus and where the resource leak occurs. I would appreciate any direction/insight. I am scratching my head trying to figure out where this is happening or where to look.

Just to clarify, I am not re-creating the `NetworkManagerSettings()` object.

```
root:~# dbus-monitor --system
Failed to open connection to system bus: The maximum number of active connections for UID 0 has been reached
```

Python code executed every 15 minutes
``` py
from sdbus_async.networkmanager import NetworkManagerSettings
self._nm_settings = NetworkManagerSettings()
try:
        # Get connection ID and see if WiFi was used as the connection medium
        conn_list = await self._nm_settings.get_connections_by_id(NetworkMngr.WIRELESS_CONNECTION_ID)

         # Verify the connection id exists, wifi is not connected, and we have no internet connection currently
         if len(conn_list) > 0 and not self.is_wifi_connected and not self.is_internet_connected:
             log.warning("Restarting [NetworkManager]...")
             await self._systemd.restart_unit("NetworkManager.service", "replace")
      except Exception as ex:
          log.exception(ex)
```

**Comment by igo95862** [comment]
Hello @wes8ty 

Are you sure the process that runs the script exits? This should close all connections.

**Comment by igo95862** [comment]
You can also use `ss --unix --all` to list all unix socket connections.

**Comment by igo95862** [comment]
@wes8ty any updates?

**Comment by wes8ty** [comment]
Hello @igo95862.

Thank you so much for the quick answer. I am still trying to dig into it but still having a lot of trouble tracking this down. If I may ask, when does the sdbus library create the connection with the dbus dameon? How is the connection created / released when using the library?

**Comment by igo95862** [comment]
If you don't explicitly pass around the bus object such as created by `sd_bus_open_system` then the default bus would be used. The default bus is thread local so every new thread will spawn a new bus which might be the source of your issues. I would recommend opening a single system bus and pass it around for each NetworkManager object initializer. This would also mean the used bus will not switch between system/session then the code is run as root or as user. 

> (never exists)

Did you mean "never exi**ts**"?

> How is the connection created / released when using the library?

The bus object can either be explicitly closed or it will be automatically closed when garbage collected. However, I have not tested what happens to a thread local variables then the thread is destroyed.

**Comment by wes8ty** [comment]
Thanks for the context! I am on the hunt for where code could be called. The program executes the following in our `__main__.py` file. Does this method allow for it to be replaced avoiding the problem? `request_default_bus_name`

``` py
# Set default bus path
sdbus.set_default_bus(sdbus.sd_bus_open_system())
```

I would like to add a logging statement when a new connection is created but I cannot find the code that performs this. Most of the methods are `NotImplemented`. Do you know of a way to accomplish this? I am hoping to get a stacktrace of when it opens a new connection.

I think it is this piece of code but wanted to be sure

``` py
def get_default_bus() -> SdBus:
    try:
        return DEFAULT_BUS.get()
    except LookupError:
        new_bus = sd_bus_open()
        DEFAULT_BUS.set(new_bus)
        return new_bus
```

Yes, I meant exit :) For those troubleshooting similar issues, the command `busctl -l` confirmed it was our process that was opening all these connections.

**Comment by igo95862** [comment]
`get_default_bus()` will be used if you don't pass the bus object explicitly to the proxy object. A new context will open a new bus every time.

**Comment by igo95862** [comment]
You can probably insert a bunch of `print` or `breakpoint` to the `get_default_bus` to see how many default busses are opened.

**Comment by wes8ty** [comment]
Thanks! I implemented the fixes you suggested and will be digging into this more with some print statements.

**Comment by igo95862** [comment]
I might switch to [thread local storage](https://docs.python.org/3/library/threading.html#thread-local-data) for default bus as context vars seems to switch context a lot.

**Comment by wes8ty** [comment]
Thanks @igo95862. I would agree to switch it to thread local storage. The `ContextVar` creates a lot of contexts.

**Comment by igo95862** [comment]
Did you managed to solve your problem?

**Comment by wes8ty** [comment]
I thought I did but it is cropping up again with the same error. Switch to thread local storage would be very nice if possible.

**Comment by igo95862** [comment]
Would you be able to test it once I create a patch for thread local storage?

**Comment by wes8ty** [comment]
Yes, I will!

**Comment by wes8ty** [comment]
Do you think it is a wise decision to create some logging when new bus is created?

**Comment by igo95862** [comment]
Makes sense to log it. Although I haven't started adding logging this might be a good spot to start.

**Comment by igo95862** [comment]
@wes8ty I added the draft commit fbf9475d11cc981c3d5b86a82fc4968ec761a6c2 which should switch default bus to thread local storage. I also added an info level log message then a new default bus is created. Can you give it a test?

**Comment by wes8ty** [comment]
Sounds good. Will test on Monday if that is ok?

**Comment by igo95862** [comment]
Sure.

**Comment by wes8ty** [comment]
It looks like it is working. I will be able to tell in time. Thanks for adding that logging statement! When do you think you will do a release?

**Comment by igo95862** [comment]
I want to do a release before Debian 13 freeze so that it will be included. There were a couple new features ideas I had like a magic proxies that create interfaces dynamically but I think I will postpone it for next release.

**Comment by igo95862** [comment]
I think I will make the default bus a hybrid of thread-local/context-local. The thread local will be the default but if the user wants they can run a context with a different default bus.

**Comment by igo95862** [comment]
@wes8ty I pushed the 4dac6ca7249a6ae570943846e562e6f51bee0fe1 which includes the same thread local bus patch. I plan on releasing an RC1 version soon followed by a full release.

**Comment by igo95862** [comment]
0.14.0 has been released with the thread-local default bus.
